core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -0,0 +1,271 @@
+                    ),
+                    QuotaFactory.UnboundedQuota$.MODULE$,
+                    responsePartitionData -> {
+                        log.trace("Data successfully retrieved by replica manager: {}", responsePartitionData);
+                        List<Tuple2<TopicIdPartition, FetchPartitionData>> responseData = CollectionConverters.asJava(
+                                responsePartitionData);
+                        processFetchResponse(shareFetchPartitionData, responseData).whenComplete(
+                                (result, throwable) -> {
+                                    if (throwable != null) {
+                                        log.error("Error processing fetch response for share partitions", throwable);
+                                        shareFetchPartitionData.future().completeExceptionally(throwable);
+                                    } else {
+                                        shareFetchPartitionData.future().complete(result);
+                                    }
+                                    // Releasing the lock to move ahead with the next request in queue.
+                                    releasePartitionsLock(shareFetchPartitionData.groupId(), topicPartitionData.keySet());
+                                });
+                        return BoxedUnit.UNIT;
+                    });
+        } catch (Exception e) {
+            // Release the locks acquired for the partitions in the share fetch request in case there is an exception
+            log.error("Error processing delayed share fetch request", e);
+            shareFetchPartitionData.future().completeExceptionally(e);
+            releasePartitionsLock(shareFetchPartitionData.groupId(), topicPartitionData.keySet());
+        }
+    }
+
+    /**
+     * Try to complete the fetch operation if we can acquire records for any partition in the share fetch request.
+     */
+    @Override
+    public boolean tryComplete() {
+        log.trace("onTry of delayed share fetch request for group {}, member {}, topic partitions {}",
+                shareFetchPartitionData.groupId(), shareFetchPartitionData.memberId(),
+                shareFetchPartitionData.partitionMaxBytes().keySet());
+
+        boolean canAnyPartitionBeAcquired = false;
+        for (TopicIdPartition topicIdPartition: shareFetchPartitionData.partitionMaxBytes().keySet()) {
+            SharePartition sharePartition = partitionCacheMap.get(new SharePartitionManager.SharePartitionKey(
+                    shareFetchPartitionData.groupId(), topicIdPartition));
+            if (sharePartition.maybeAcquireFetchLock()) {
+                if (sharePartition.canAcquireRecords()) {
+                    canAnyPartitionBeAcquired = true;
+                }
+                sharePartition.releaseFetchLock();
+                if (canAnyPartitionBeAcquired)
```

Comment:
1. This is not very accurate. `canAnyPartitionBeAcquired` doesn't mean that there is enough bytes to satisfy the minByte requirement for the request. We need to take into consideration the size of all available batches in SharePartition.
2. There is also a synchronization issue. In DelayedFetch, there is only 1 fetch request per partition per group. So, after checking that a particular fetch request has enough bytes, we can by sure that when `onComplete` is called, there is indeed enough bytes. In DelayedShareFetch, there could be multiple share fetch requests per partition per group. What could happen is that in one thread, we check that request 1 can be completed, but by the time we try to complete the request, another thread has given the data in that partition to request 2. This means that we may return fewer bytes for request 1.

Reply:
Hi @junrao,
1. You're right in pointing that out. In this initial version of purgatory, the only constraint that I've included is delaying a request in case the record lock partition limit is reached. We will be targeting to have other constraints like `minBytes` in the future PRs. I have updated the PR description to highlight this fact to avoid confusion.
2. I think you're right about the scenario. I am not sure if it is a problem if request 2 gets the data and request 1 contains an empty response (since now we're using `replicaManager.readFromLog` instead of `replicaManager.fetchMessages`). Please if you can elaborate the problem, that would be really helpful.
cc - @mumrah @AndrewJSchofield

Reply:
>  I am not sure if it is a problem if request 2 gets the data and request 1 contains an empty response (since now we're using replicaManager.readFromLog instead of replicaManager.fetchMessages). 

If we return request 1 with an empty response before maxWait, it means that we are returning the response too early, right? The response is supposed to be sent when either minByte or maxWait is satisfied and minByte is always > 0.

Reply:
Hi @junrao , thanks for that explanation. I have modified my code in a way that now if a request can be completed in `tryComplete`, I do not release the share partition's lock and prepare my topic partition data within the variable `topicPartitionDataFromTryComplete`. I use this variable within `onComplete` in case the call to `onComplete` came via `tryComplete`. Thus the partitions are locked once we see that there is a chance `tryComplete` can succeed. I think this should resolve the synchronization issue. Thanks.

Reply:
Also, there are still chances that occasionally there might be a response sent back to the client where response bytes is lesser than `minBytes` and `MaxWaitMs` hasn't happened yet. Discussed this with @AndrewJSchofield , he is fine with that in this PR since here we want to ensure that ShareFetch requests do not immediately return if they hit the partition lock limit. Once we have `minBytes` criteria covered in future PRs, this problem should go away.